### PR TITLE
Replace cable_ready responses with turbo_stream OFN-12794

### LINKED ISSUE #12794: Replace mrujs with turbo
### What we should change and why (this is tech debt)

We currently use mrujs and turbo-rails together. While they can co-exist, Turbo should be able to do everything that UJS used to do. We may need to adjust some files though.

* [Upgrading from Rails UJS / Turbolinks to Turbo](https://github.com/hotwired/turbo-rails/blob/main/UPGRADING.md#upgrading-from-rails-ujs--turbolinks-to-turbo) - :warning: We are not upgrading from rails-ujs, but from mrujs.
* HTML attributes may need renaming, like `data-disable-with` to `data-turbo-disable-with`. Verify the need for this first!
* Replace event handlers like `ajax:beforeSend` if needed.

### Context

* https://github.com/openfoodfoundation/openfoodnetwork/pull/12788

### PR DESCRIPTION
**Pull request by wandji20**
#### What? Why?

- Closes #12794

   Since Turbo already covers everything cable_ready does, we can safely move from cable_ready and mrujs to turbo.

#### What should we test?
The overall functionality of the app should be retained.

#### Release notes

Replace mrujs with turbo

Changelog Category (reviewers may add a label for the release notes):

- [x] Technical changes only

Replace mrujs with turbo

#### Dependencies
N/A

#### Documentation updates
N/A

### CODE REVIEW

File: spec/system/admin/order_spec.rb [hunk 1]
```diff
@@ -951,10 +951,12 @@ def new_order_with_distribution(distributor, order_cycle)
           order.finalize! # ensure order has a payment to capture
           order.payments << create(:check_payment, order:, amount: order.total)
           order.payments.first.capture!
+          login_as_admin
```
Comment (wandji20) [inline spec/system/admin/order_spec.rb:954]: I could not figure out why the spec `@enterprise_user` had no permission to perform order ship action.

File: app/views/layouts/_alert.html.haml [hunk 2]
```diff
@@ -1,5 +1,6 @@
 .alert-box{ class: "#{type}" }
   = message
   - if local_assigns[:unconfirmed]
-    %a{ "data-action": "login-modal#resend_confirmation", "data-tab": local_assigns[:tab] }
```
Comment (rioug) [inline app/views/layouts/_alert.html.haml:4]: I believe we can now remove `resend_confirmation`, could you double check ?

File: spec/system/admin/order_spec.rb [hunk 3]
```diff
@@ -951,10 +951,12 @@ def new_order_with_distribution(distributor, order_cycle)
...
           visit spree.edit_admin_order_path(order)
         end
 
         it "ships the order and shipment email is sent" do
+          login_as_admin
```
Comment (rioug) [inline spec/system/admin/order_spec.rb:959]: Why is this needed ? it should already be logged as an admin because it's done in the `before` block.

File: spec/system/admin/order_spec.rb [hunk 4]
```diff
@@ -972,6 +974,7 @@ def new_order_with_distribution(distributor, order_cycle)
         end
 
         it "ships the order without sending email" do
+          login_as_admin
```
Comment (rioug) [inline spec/system/admin/order_spec.rb:977]: Same as above

File: app/views/spree/admin/orders/index.html.haml [hunk 5]
```diff
@@ -21,14 +21,36 @@
...
+    .margin-bottom-30
+      = t("js.admin.orders.cancel_the_order_html")
+    .margin-bottom-30
+      = render partial: "spree/admin/orders/messages/cancel_orders"
+
+    %p.modal-actions.justify-space-around
+      %button.button.secondary{ type: "button", 'data-action': 'click->modal#close' }
+        = t('js.admin.modals.cancel')
+      %button.button.primary{ type: 'submit' }
+        = t('js.admin.modals.confirm')
```
Comment (rioug) [inline app/views/spree/admin/orders/index.html.haml:56]: Any reasons why you are not using `ConfirmModalComponent` here ?

File: app/webpacker/controllers/bulk_actions_controller.js [hunk 6]
```diff
@@ -1,24 +1,43 @@
...
+        "Content-Type": "application/json",
+        "X-CSRF-Token": csrfToken
+      },
+      body: JSON.stringify(data),
+    })
+      .then((response) => response.text())
+      .then((html) => {
+        Turbo.renderStreamMessage(html);
+      })
+      .catch((error) => console.error(error));
```
Comment (rioug) [inline app/webpacker/controllers/bulk_actions_controller.js:33]: Maybe I am missing something here, but it feels like we are replicating something that `Turbo` should be doing for us ? Is there a way to refactor this so we don't have to manually use `fetch` ?

**Review by rioug — COMMENTED** [review]
Great work, that's awesome @wandji20 ! I just have a couple of questions, see my comments.

File: app/views/layouts/_alert.html.haml (hunk 2 above)
Comment (wandji20) [reply app/views/layouts/_alert.html.haml:4]: It was removed in a later commit.

File: spec/system/admin/order_spec.rb (hunk 3 above)
Comment (wandji20) [reply spec/system/admin/order_spec.rb:959]: These specs are in the context of an enterprise manager and we have an enterprise manager signed in this block 
```
    before(:each) do
      @enterprise_user = create(:user)
      @enterprise_user.enterprise_roles.build(enterprise: supplier1).save
      @enterprise_user.enterprise_roles.build(enterprise: coordinator1).save
      @enterprise_user.enterprise_roles.build(enterprise: distributor1).save

      login_as @enterprise_user
    end
    ```
    I had some trouble figuring out why the permission failed for the signed in enterprise user when trying to ship an order.
    So I was hoping to get help during review.

File: app/views/spree/admin/orders/index.html.haml (hunk 5 above)
Comment (wandji20) [reply app/views/spree/admin/orders/index.html.haml:56]: I did this simply because the confirm modal did not support handling forms without refactoring.
I will refactor to remove these duplications.

File: app/webpacker/controllers/bulk_actions_controller.js (hunk 6 above)
Comment (rioug) [reply app/webpacker/controllers/bulk_actions_controller.js:33]: I see, maybe a comment would be useful here.

**Review by rioug — CHANGES_REQUESTED** [review]
Thanks for the refactoring it looks cleaner :pray: . There are still a couple of things to address before we can merge this.

**Review by rioug — APPROVED** [review]
Awesome ! Thanks @wandji20 it looks good now :+1:

File: app/controllers/spree/users_controller.rb [hunk 7]
```diff
@@ -31,13 +30,10 @@ def registered_email
...
-        render status: :ok, cable_ready: cable_car.
-          inner_html(
-            "#login-feedback",
-            partial("layouts/alert",
-                    locals: { type: "alert", message: t('devise.failure.already_registered') })
-          ).
-          dispatch_event(name: "login:modal:open")
+        respond_to do |format|
+          format.html { head :ok }
+          format.turbo_stream { :registered_email }
```
Comment (dacook) [inline app/controllers/spree/users_controller.rb:35]: This change isn't accompanied with a change in users_controller_spec. We need to add `format: :turbo_stream`.

I was expecting it to fail, because you forgot to write `render:` before `:registered_email`. But Rails' default handling kicks in and renders it anyway ;) 
Because it's simply rendering a partial of the same name, we don't need the block at all:

```suggestion
          format.turbo_stream
```

Comment (dacook) [reply app/controllers/spree/users_controller.rb:35]: Hmm, this isn't covered by a system spec. 
In fact, on closer look, I can see it's not even used!  After some git detective work, I managed to find out it hasn't been used for over a year, since https://github.com/openfoodfoundation/openfoodnetwork/commit/fc0cca2210dcd78f4554fea0dc76af712c05922c#diff-96984374385f56a535a566ac8afee14de6461121ed5717e2ddced0e640d2c75b 

So we can remove this action entirely!

File: app/controllers/spree/users_controller.rb [hunk 8]
```diff
@@ -48,12 +44,12 @@ def create
...
+        redirect_to main_app.root_path
       else
-        render status: :unprocessable_entity, cable_ready: cable_car.morph(
-          "#signup-tab",
-          partial("layouts/signup_tab", locals: { signup_form_user: @user })
-        )
+        respond_to do |format|
+          format.html { head :unprocessable_entity }
+          format.turbo_stream { render :create, status: :unprocessable_entity }
+        end
```
Comment (dacook) [inline app/controllers/spree/users_controller.rb:52]: This change isn't covered by the controller spec. I _think_ it is covered by this spec: https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/system/consumer/authentication_spec.rb#L90-L95

I can't verify right now, as there's something wrong with my dev environment 🫠 
Would you be able to double-check this? I think you should be able to see it if you run Chrome in non-headless mode like this:
```
HEADLESS=false bundle exec rspec spec/system/consumer/authentication_spec.rb:85
```

**Review by dacook — COMMENTED** [review]
I've partially reviewed up to "Handle user registration responses with turbo_stream".
I have a few comments, would you be able to take a look?

File: app/controllers/spree/users_controller.rb (hunk 7 above)
Comment (wandji20) [reply app/controllers/spree/users_controller.rb]: Thanks
Nice one spotting this one

File: app/controllers/spree/user_sessions_controller.rb [hunk 9]
```diff
@@ -60,11 +60,19 @@ def email_unconfirmed?
...
 
     def render_unconfirmed_response
-      render status: :unprocessable_entity, cable_ready: cable_car.inner_html(
-        "#login-feedback",
-        partial("layouts/alert", locals: { type: "alert", message: t(:email_unconfirmed),
-                                           unconfirmed: true, tab: "login" })
-      )
+      message = t(:email_unconfirmed)
+      local_values = { type: "alert", message:, unconfirmed: true,
+                       tab: "login", email: params.dig(:spree_user, :email) }
```
[the remark this answers is not in the document]
Comment (wandji20) [reply app/controllers/spree/user_sessions_controller.rb:65]: The email was used to resend confirmation via stimulus with login_modal#resend_confirmation.
I will add a commit to clean up setting the email in the login_modal_cotroller.js as it is no longer required.

File: app/controllers/spree/users_controller.rb (hunk 8 above)
Comment (wandji20) [reply app/controllers/spree/users_controller.rb:38]: That's correct, failed user signup spec is handled in `spec/system/consumer/authentication_spec.rb:85`

File: app/webpacker/controllers/turbo_search_controller.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    const form = e.target;
+    const url = new URL(form.action);
+    const formData = new FormData(form);
+    const params = new URLSearchParams(formData).toString();
+    
+    // Manually visit the new URL with the search params
+    // inorder to preserve params
+    Turbo.visit(`${url.pathname}?${params}`, { action: 'replace' });
+  }
+}
```
Comment (dacook) [inline app/webpacker/controllers/turbo_search_controller.js:25]: Nice work, but this looks like something that Turbo should be able to handle for us.

Does this work instead? https://stackoverflow.com/a/76283733/421243

> 
> For GET forms and links you have to specify that you want to send a TURBO_STREAM request:
> ```
> <%= form_with url: search_orders_path, method: :get,
>    data: {turbo_stream: true} do |f| %>
> ...
> <% end %> 
> ```

File: app/webpacker/channels/session_channel.js [hunk 11]
```diff
@@ -1,8 +1,9 @@
 import consumer from './consumer'
-import CableReady from 'cable_ready'
 
 consumer.subscriptions.create("SessionChannel", {
   received(data) {
-    if (data.cableReady) CableReady.perform(data.operations)
+    if (!data.selector) return;
+
+    document.querySelector(data.selector).innerHTML = data.html;
   }
```
Comment (dacook) [inline app/webpacker/channels/session_channel.js:8]: I don't think there's a reason to replace CableReady with ActionCable. CableReady is adding a feature, meaning we can use less code.
Remember that the original issue was simply to replace mrujs, not to replace CableReady

`±` Can you please remove/revert this, and the similar commits? 
* Replace bulk invoice CableReady broadcast with ActionCable
* Replace admin report CableReady broadcast with ActionCable

Comment (dacook) [reply app/webpacker/channels/session_channel.js:8]: Ok, I see that @mkllnk mentioned ActionCable [here](https://github.com/openfoodfoundation/openfoodnetwork/issues/12794#issuecomment-2370000245). I'm not sure if this is what he intended though.

Maikel, do you have an opinion if this change is a good idea or not? There's a couple more commits following which give more context.

**Review by dacook — CHANGES_REQUESTED** [review]
Thanks for your patience as I've been sick and also busy. This is huge undertaking and is coming along well. Changing CableReady to TurboStreams should be more reliable and hopefully easier to maintain! 
I have a suggestion to simplify, and will await clarification on another comment.

I can also see now that this is covering _a lot_ of the codebase, and I'm thinking we may need to consider dividing into multiple PRs so that we can test it carefully and release it in parts. 
Looking at the changes, I think I'd suggest something like three chunks:
* User login and profile
* Checkout
* Admin

But I'm not sure how easy it would be to split up.. will discuss with the team.

**Comment by wandji20** [comment]
I agree splitting it into separate PRs, @dacook will be better.
So I will start with the User sections and later close this PR.

**Comment by dacook** [comment]
Awesome, thank you Wandji!
